File: parcel-locker-cpp/library/src/model/Locker.cpp
//
// Created by student on 31.05.2021.
//

#include "model/Locker.h"
#include "model/DepositBox.h"
#include "exceptions/LockerException.h"


Locker::Locker(int boxAmount)
{
    try {
        if (!boxAmount)
            throw LockerException("Locker with 0 boxes can not be created!");
    }
    catch(LockerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }

    for(int i = 0; i < boxAmount; i++){
        depositBoxes.push_back(std::make_shared<DepositBox>(std::to_string(i)));
    }
}

void Locker::putIn(const boost::uuids::uuid &id, const std::string& telNumber, const std::string& accessCode) {
    for(int i = 0; i < depositBoxes.size(); i++) {
        if(!depositBoxes.at(i)->getStatus()) {
            depositBoxes.at(i)->setAccessCode(accessCode);
            depositBoxes.at(i)->setStatus(true);
            depositBoxes.at(i)->setTelNumber(telNumber);
            depositBoxes.at(i)->setDeliveryId(id);
            return;
        }
    }
    try {
        throw LockerException("Not able to put package into box.");
    }
    catch (LockerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }
}

boost::uuids::uuid Locker::takeOut(const std::string &telNumber, const std::string &code)
{
    for(int i = 0; i < depositBoxes.size(); i++) {
        if(depositBoxes.at(i)->Access(code,telNumber)) {   ///korzystamy tu z metody depositBoxa, która sama czyści
            return depositBoxes.at(i)->getDeliveryId();
        }
    }
    try {
        throw LockerException("Couldn't get package out.");
    }
    catch(LockerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }
}

std::string Locker::getInfo()
{
    std::string lista = "Empty boxes: " + std::to_string(this->findAllEmpty()) + "\n";
    for(int i=0; i<depositBoxes.size();i++)
    {
        if(!depositBoxes.at(i)->getStatus())
        {
            lista += depositBoxes.at(i)->getId() + "\n";
        }
    }
    return lista;
}

int Locker::findAllEmpty() {
    int counter = 0;
    for(int i = 0; i < depositBoxes.size(); i++) {
        if (!depositBoxes.at(i)->getStatus()) {
            counter++;
        }
    }
    return counter;
}

DepositBoxPtr Locker::getDepositBox(std::string nrId) {
    for(int i=0; i<depositBoxes.size();i++)
    {
        if(depositBoxes.at(i)->getId() == nrId)
        {
            return depositBoxes.at(i);
        }
    }
}

File: parcel-locker/src/main/java/pl/pas/parcellocker/model/locker/Locker.java
package pl.pas.parcellocker.model.locker;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.OneToMany;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import pl.pas.parcellocker.exceptions.LockerException;
import pl.pas.parcellocker.model.EntityModel;
import pl.pas.parcellocker.model.delivery.Delivery;


@Entity
@NoArgsConstructor
@EqualsAndHashCode
public class Locker extends EntityModel {

    private String identityNumber;
    private String address;

    @OneToMany(fetch = FetchType.LAZY, cascade = {CascadeType.PERSIST, CascadeType.REMOVE, CascadeType.MERGE})
    private List<DepositBox> depositBoxes;

    public Locker(String identityNumber, String address, int boxAmount) {
        try {
            if (boxAmount <= 0)
                throw new LockerException("Locker with 0 boxes can not be created!");
        } catch (LockerException e) {
            //log.error(e.getMessage());
        }

        this.identityNumber = identityNumber;
        this.address = address;
        depositBoxes = new ArrayList<>();
        for (int i = 0; i < boxAmount; i++) {
            depositBoxes.add(new DepositBox());
        }
    }

    public UUID putIn(Delivery delivery, String telNumber, String accessCode) {

        for (DepositBox depositBox : depositBoxes) {
            if (depositBox.isEmpty()) {
                depositBox.putIn(delivery, telNumber, accessCode);
                return depositBox.getId();
            }
        }
        throw new LockerException("Not able to put package with id = " +
            delivery.getId() + " into locker " + this.getIdentityNumber() + ".");
    }

    public Delivery takeOut(String telNumber, String code) {
        for (DepositBox depositBox : depositBoxes) {
            if (depositBox.canAccess(code, telNumber)) {
                Delivery takenOutDelivery = depositBox.getDelivery();
                depositBox.clean();
                return takenOutDelivery;
            }
        }
        throw new LockerException("Couldn't get any package out with access code: "
            + code
            + "and phone number: "
            + telNumber);
    }

    public int countEmpty() {
        int counter = 0;
        for (DepositBox depositBox : depositBoxes) {
            if (depositBox.isEmpty()) {
                counter++;
            }
        }
        return counter;
    }

    public DepositBox getDepositBox(UUID id) {
        for (DepositBox depositBox : depositBoxes) {
            if (depositBox.getId().equals(id)) {
                return depositBox;
            }
        }
        return null;
    }

    public String getIdentityNumber() {
        return identityNumber;
    }

    public String getAddress() {
        return address;
    }
}
